### PR TITLE
Add raytrace outside the map

### PR DESCRIPTION
**Pull request by renan028**
# Description
[AB#17690](https://dev.azure.com/rapyuta-robotics/flappter/_workitems/edit/17690)
Add the option to raytrace free space even if sensor is outside the map

https://github.com/user-attachments/assets/da8e5269-d2f9-4618-9b59-6277bc414067

### CODE REVIEW

File: costmap_2d/cfg/ObstaclePlugin.cfg [hunk 1]
```diff
@@ -7,6 +7,7 @@ gen = ParameterGenerator()
 gen.add("enabled", bool_t, 0, "Whether to apply this plugin or not", True)
 gen.add("footprint_clearing_enabled", bool_t, 0, "Whether to clear the robot's footprint of lethal obstacles", True)
 gen.add("max_obstacle_height", double_t, 0, "The maximum height of any obstacle to be inserted into the costmap in meters.", 2, 0, 50)
+gen.add("raytrace_outside_map", bool_t, 0, "Whether to raytrace when sensor is outside map or not", False)
```
Comment (corot) [inline costmap_2d/cfg/ObstaclePlugin.cfg]:
```suggestion
gen.add("raytrace_outside_map", bool_t, 0, "Whether to raytrace when the sensor is outside the map", False)
```

File: costmap_2d/plugins/obstacle_layer.cpp [hunk 2]
```diff
@@ -495,8 +511,10 @@ void ObstacleLayer::raytraceFreespace(const Observation& clearing_observation, d
   const sensor_msgs::PointCloud2 &cloud = *(clearing_observation.cloud_);
 
   // get the map coordinates of the origin of the sensor
+
   unsigned int x0, y0;
-  if (!worldToMap(ox, oy, x0, y0))
+  const bool origin_valid = worldToMap(ox, oy, x0, y0);
+  if (!raytrace_outside_map_ && !origin_valid)
```
Comment (corot) [inline costmap_2d/plugins/obstacle_layer.cpp]: S: reads more logic
```suggestion
  if (!origin_valid && !raytrace_outside_map_)
```

File: costmap_2d/plugins/obstacle_layer.cpp [hunk 3]
```diff
@@ -570,6 +592,41 @@ void ObstacleLayer::raytraceFreespace(const Observation& clearing_observation, d
...
+  // Define the sensor ray as a linestring (from the sensor origin to the endpoint)
+  Linestring sensor_ray;
+  bg::append(sensor_ray, Point(ox, oy));
+  bg::append(sensor_ray, Point(wx, wy));
+
+  std::vector<Point> intersection_points;
+  bg::intersection(sensor_ray, map_boundary_, intersection_points);
+
+  if (intersection_points.size() != 2)
+  {
```
Comment (corot) [inline costmap_2d/plugins/obstacle_layer.cpp:612]: S: explain what happens here, and also on the calling place

File: costmap_2d/include/costmap_2d/obstacle_layer.h [hunk 4]
```diff
@@ -141,12 +153,27 @@ class ObstacleLayer : public CostmapLayer
   virtual void raytraceFreespace(const costmap_2d::Observation& clearing_observation, double* min_x, double* min_y,
                                  double* max_x, double* max_y);
 
+  /**
```
Comment (corot) [inline costmap_2d/include/costmap_2d/obstacle_layer.h:156]: S: mentions somewhere "if raytrace_outside_map is true"

**Review by corot — APPROVED** [review]

**Review by tomin11 — APPROVED** [review]

**Pull request merged**